cpu/kinetis_common/periph/gpio.c

```diff
@@ -24,384 +24,228 @@
- * applications. */
-static gpio_int_config_entry_t* gpio_interrupts[PORT_NUMOF];
-
-static mutex_t int_config_lock = MUTEX_INIT;
-
-/* Maximum number of simultaneously enabled GPIO interrupts. Each pool entry
- * uses 20 bytes of RAM.
+static isr_ctx_t isr_ctx[CTX_NUMOF] = {{0}};
+
+/**
+ * @brief Allocation of 4 bit per pin to map a pin to an interrupt context
  */
-#ifndef GPIO_INT_POOL_SIZE
-#define GPIO_INT_POOL_SIZE 16
+static uint32_t isr_map[] = {
+    0xffffffff, 0xffffffff, 0xffffffff, 0xffffffff,
+    0xffffffff, 0xffffffff, 0xffffffff, 0xffffffff,
+    0xffffffff, 0xffffffff, 0xffffffff, 0xffffffff,
+    0xffffffff, 0xffffffff, 0xffffffff, 0xffffffff,
+#if GPIO_PORTS_NUMOF > 4
+    0xffffffff, 0xffffffff, 0xffffffff, 0xffffffff,
+#endif
+#if GPIO_PORTS_NUMOF > 5
+    0xffffffff, 0xffffffff, 0xffffffff, 0xffffffff,
+#endif
+#if GPIO_PORTS_NUMOF > 6
+    0xffffffff, 0xffffffff, 0xffffffff, 0xffffffff,
+#endif
+#if GPIO_PORTS_NUMOF > 7
+    0xffffffff, 0xffffffff, 0xffffffff, 0xffffffff
 #endif
+};
 
-/* Pool of linked list entries which can be used by any port configuration.
- * Rationale: Avoid dynamic memory inside low level periph drivers. */
-gpio_int_config_entry_t config_pool[GPIO_INT_POOL_SIZE];
 
-static PORT_Type * const _port_ptrs[] = PORT_BASE_PTRS;
-static GPIO_Type * const _gpio_ptrs[] = GPIO_BASE_PTRS;
+static inline PORT_Type *port(gpio_t pin)
+{
+    return (PORT_Type *)(PORT_ADDR_BASE | (pin & PORT_ADDR_MASK));
+}
 
-static inline uint32_t _port_num(gpio_t dev) {
-    return (uint32_t)((dev & GPIO_PORT_MASK) >> GPIO_PORT_SHIFT);
+static inline GPIO_Type *gpio(gpio_t pin)
+{
+    return (GPIO_Type *)(GPIO_ADDR_BASE | (pin & GPIO_ADDR_MASK));
 }
 
-static inline uint8_t _pin_num(gpio_t dev) {
-    return (uint8_t)((dev & GPIO_PIN_MASK) >> GPIO_PIN_SHIFT);
+static inline int port_num(gpio_t pin)
+{
+    return (int)((pin >> GPIO_SHIFT) & 0x7);
 }
 
-static inline PORT_Type *_port(gpio_t dev) {
-    return _port_ptrs[_port_num(dev)];
+static inline int pin_num(gpio_t pin)
+{
+    return (int)(pin & 0x3f);
 }
 
-static inline GPIO_Type *_gpio(gpio_t dev) {
-    return _gpio_ptrs[_port_num(dev)];
+static inline void clk_en(gpio_t pin)
+{
+    BITBAND_REG32(SIM->SCGC5, SIM_SCGC5_PORTA_SHIFT + port_num(pin)) = 1;
 }
 
-static void _clear_interrupt_config(gpio_t dev) {
-    gpio_int_config_entry_t* entry = NULL;
-    uint8_t pin_number = _pin_num(dev);
-    mutex_lock(&int_config_lock);
-    /* Search for the given pin in the port's interrupt configuration */
-    LL_SEARCH_SCALAR(gpio_interrupts[_port_num(dev)], entry, pin, pin_number);
-    if (entry != NULL) {
-        LL_DELETE(gpio_interrupts[_port_num(dev)], entry);
-        /* pin == 0 means the entry is available */
-        entry->pin = 0;
-    }
-    mutex_unlock(&int_config_lock);
+/**
+ * @brief Get context for a specific pin
+ */
+static inline int get_ctx(int port, int pin)
+{
+    return (isr_map[(port * 4) + (pin >> 3)] >> ((pin & 0x7) * 4)) & 0xf;
```

Comment:
Sure: the concept is actually rather simple: we allocate an array with a fixed (maximum) number of interrupt contexts. Then we allocate for each MCU pin a number bit, that are pointers into the isr-context array. 

Now we use 4-bit per GPIO pin (allocated in `isr_map[]`). This allows us to 'address' 16 different positions in the 'isr_ctx[]' array -> so we can have up to 16 pins configured concurrently as interrupt source (-> `CTX_NUMOF < 16`).

The line above now extracts the 'isr_ctx[]' offest for a given port/pin combination. The first part finds the word in the `isr_map[]` array which contains the 4-bit value for the given pin